webapp/src/main/java/com/box/l10n/mojito/smartling/SmartlingOAuth2TokenService.java

```diff
@@ -0,0 +1,167 @@
+  }
+
+  private boolean isTokenExpired() {
+    return smartlingOAuthAccessToken == null
+        || smartlingOAuthAccessToken.getTokenExpiryTime() <= System.currentTimeMillis() + 3000;
+  }
+
+  private boolean isRefreshTokenExpired() {
+    return smartlingOAuthAccessToken == null
+        || smartlingOAuthAccessToken.getRefreshExpiryTime() <= System.currentTimeMillis() + 3000;
+  }
+
+  private boolean isRefreshRequired() {
+    // check if the access or refresh tokens are about to expire
+    return isTokenExpired() || isRefreshTokenExpired();
+  }
+
+  private synchronized SmartlingOAuthAccessResponse.TokenData requestAccessToken() {
+    if (isRefreshTokenExpired()) {
+      requestNewAccessToken();
+    } else if (isTokenExpired()) {
+      refreshAccessToken();
+    }
+
+    return smartlingOAuthAccessToken;
+  }
+
+  private void requestNewAccessToken() {
+    ClientAuthDetails authDetails = new ClientAuthDetails(clientId, clientSecret);
+    try {
+      String json = objectMapper.writeValueAsString(authDetails);
+      retrieveToken(accessTokenUrl, json);
+    } catch (URISyntaxException | IOException | InterruptedException e) {
+      logger.error("Exception occurred retrieving OAuth token from Smartling: ", e);
+      throw new SmartlingOAuthTokenException(
+          "Exception occurred retrieving OAuth token from Smartling: ", e);
+    }
+  }
+
+  private void refreshAccessToken() {
+    RefreshRequest refreshRequest = new RefreshRequest(smartlingOAuthAccessToken.getRefreshToken());
+    try {
+      String json = objectMapper.writeValueAsString(refreshRequest);
+      retrieveToken(refreshTokenUrl, json);
+    } catch (URISyntaxException | IOException | InterruptedException e) {
+      logger.error("Exception occurred refreshing OAuth token with Smartling: ", e);
+      throw new SmartlingOAuthTokenException(
+          "Exception occurred refreshing OAuth token with Smartling: ", e);
+    }
+  }
+
+  private void retrieveToken(String authUrl, String json)
+      throws URISyntaxException, IOException, InterruptedException {
+    HttpRequest request =
+        HttpRequest.newBuilder()
+            .uri(new URI(authUrl))
+            .header("Content-Type", "application/json")
+            .POST(HttpRequest.BodyPublishers.ofString(json))
+            .build();
+
+    HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());
+
+    SmartlingOAuthAccessResponse smartlingOAuthAccessResponse =
+        objectMapper.readValue(response.body(), SmartlingOAuthAccessResponse.class);
+    smartlingOAuthAccessToken = smartlingOAuthAccessResponse.getResponse().getData();
+    smartlingOAuthAccessToken.setTokenExpiryTime(
+        System.currentTimeMillis() + (smartlingOAuthAccessToken.getExpiresIn() * 1000));
+    smartlingOAuthAccessToken.setRefreshExpiryTime(
+        System.currentTimeMillis() + (smartlingOAuthAccessToken.getRefreshExpiresIn() * 1000));
+  }
+
+  static class ClientAuthDetails {
```

Comment:
you can use records now, it will be way more concise :) 

something like
```
record ClientAuthDetails(@JsonProperty("userIdentifier") String clientId, @JsonProperty("userSecret") String clientSecret) {} ```